fullnode/src/main/java/io/coti/fullnode/services/TransactionService.java

```diff
@@ -631,16 +650,138 @@ private void updateAddressTransactionsByAttachment() {
+            transactionHashesByAttachmentMap.put(transactionData.getAttachmentTime(), transactionHashSet);
+            addressToTransactionsByAttachmentMap.put(baseTransactionData.getAddressHash(), transactionHashesByAttachmentMap);
+        });
+    }
+
     @Override
     public void removeTransactionHashFromUnconfirmed(TransactionData transactionData) {
         transactionPropagationCheckService.removeTransactionHashFromUnconfirmed(transactionData.getHash());
     }
 
+    @Override
+    public void handlePropagatedRejectedTransaction(RejectedTransactionData rejectedTransactionData) {
+        if (!baseNodeEventService.eventHappened(Event.TRUST_SCORE_CONSENSUS)) {
+            log.error("Error encountered during the handling of rejected transaction data, event TRUST_SCORE_CONSENSUS didn't happen");
+            return;
+        }
+        Hash rejectedTransactionHash = rejectedTransactionData.getHash();
+        if (rejectedTransactionHash == null) {
+            log.error("Error encountered during the handling of rejected transaction data, hash is null");
+            return;
+        }
+        try {
+            transactionHandlingReadWriteLock.writeLock().lock();
+            synchronized (transactionLockData.addLockToLockMap(rejectedTransactionHash)) {
+                if (!transactionHelper.isTransactionHashExists(rejectedTransactionHash)) {
+                    return;
+                }
+                if (!validationService.validatePropagatedRejectedTransactionDataIntegrity(rejectedTransactionData)) {
+                    log.error("Data Integrity validation failed for rejected transaction: {}", rejectedTransactionHash);
+                    return;
+                }
+
+                rejectTransaction(rejectedTransactionData);
+            }
+        } catch (Exception e) {
+            log.error(e.getMessage());
+        } finally {
+            transactionLockData.removeLockFromLocksMap(rejectedTransactionHash);
+            transactionHandlingReadWriteLock.writeLock().unlock();
+        }
+    }
+
+    private void rejectTransaction(RejectedTransactionData rejectedTransactionData) {
+        Hash rejectedTransactionDataHash = rejectedTransactionData.getHash();
+        TransactionData transactionData = transactions.getByHash(rejectedTransactionDataHash);
+        if (transactionData == null) {
+            throw new TransactionException("No TransactionData found for rejected transaction hash: " + rejectedTransactionDataHash);
+        }
+
+        transactionData.getChildrenTransactionHashes().forEach(hash ->
+                rejectTransaction(new RejectedTransactionData(transactions.getByHash(hash), "Transaction rejected as parent already rejected")));
+
+        log.debug("Starting to remove the rejected transaction {}", rejectedTransactionDataHash);
+        removeDataFromMemory(transactionData);
+        transactionHelper.continueHandleRejectedTransaction(transactionData);
```

Comment:
why webSocketSender.notifyTransactionHistoryChange and rejectedTransactions.put and updateAddressToRejectedTransactionsMap are not in the continueHandleRejectedTransaction?